workflow/executor/agent.go

```diff
@@ -0,0 +1,144 @@
+package executor
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+
+	log "github.com/sirupsen/logrus"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	runtimeutil "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/argoproj/argo-workflows/v3/errors"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	workflow "github.com/argoproj/argo-workflows/v3/pkg/client/clientset/versioned"
+	"github.com/argoproj/argo-workflows/v3/util"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	argohttp "github.com/argoproj/argo-workflows/v3/workflow/executor/http"
+)
+
+type AgentExecutor struct {
+	WorkflowName      string
+	ClientSet         kubernetes.Interface
+	WorkflowInterface workflow.Interface
+	RESTClient        rest.Interface
+	Namespace         string
+	CompleteTask      map[string]struct{}
```

Comment:
I don't think we should have this "CompleteTask" field - it's state that will become incorrect - use a variable instead.

Reply:
Introduced this variable to keep track the executed task and avoid agent to execute again and update the noderesult. Some scenario, controller is removing status and agent is updating again. it is going out of sync.

Reply:
ok, maybe a good idea then